### ISSUE TITLE
Background Audio Playback

### ISSUE THREAD

**Issue by anthonylavado** [jellyfin/jellyfin-ios#19]
According to some pages in Expo's documentation, this isn't supported, yet on other pages, it says it is?

Here in limitations it says no:
https://docs.expo.io/versions/latest/introduction/why-not-expo/#limitations-of-the-managed-workflow

But here it says we can do it with the right app.json entry and object:
https://docs.expo.io/versions/latest/sdk/audio/#audiosetaudiomodeasyncmode

We should see what we can do to get this working.

**Comment by Camotoy** [comment]
Since we're SO CLOSE to getting a near-perfect Jellyfin audio experience I went ahead and saw what I could do with this. I'm familiar with programming but haven't touched any form of JavaScript since an HTML class.
The `Audio.setAudioModeAsync()` function seems to be what we need, but I tried just putting it into `App.js` with no luck. And it's the function being missed entirely - I tried disabling the function's `playsInSilentModeIOS` parameter and sound still played on my iOS device on silent. My current theory is that the code only affects any local code, and this repo gets most of the UI from the web. Or, the code to show the web UI in this repository is somewhere else. 
I'm going to keep looking and testing.

**Comment by anthonylavado** [comment]
If I'm not mistaken, the Web UI in this is literally the interface from the server. That is to say, it is more like a special browser itself. Could be wrong. Paging @thornbill.

**Comment by samuel9554** [comment]
Hey guys on Ipad 5th Gen running iOS 12.4.1 I don't have this issue closing the screen, changing apps the music will continue to play. But the next song in the play list will not start. Hope it help.

**Comment by thornbill** [comment]
I believe this is caused by the js execution in the webview being stopped when in the background. I saw the same behavior in iOS 12 where it would complete the current song, but not continue to the next song. It seems that iOS 13 is much more aggressive with killing background tasks and so it does not even complete playing the current song.

I think the best solution is to use the NativeShell interface in web to inject a plugin to play audio via a native element. Hopefully that will resolve this issue. :crossed_fingers:

**Comment by RealDyllon** [comment]
It’s a known issue with expo: https://github.com/expo/expo/issues/5472

When an expo app starts playing audio, then moves to the background, the audio will continue till the end. However, the next audio file won’t start playing (on iOS). 

I think react native supports background audio playback, but that would require ejecting the entire project. 

Looks like expo is working on a fix, no ETA: https://expo.canny.io/feature-requests/p/audio-playback-in-background

**Comment by spdaly** [comment]
+1

**Comment by dannymichel** [comment]
Is the current iPhone app available in the app store outside of expo still considered the expo app? Is that why background playback isn't supported in that version either?

**Comment by thornbill** [comment]
@dannymichel yes it’s the same app.

**Comment by TwentyH** [comment]
Expo issue linked above doesn’t look like it will ever be fixed. Is there an alternative?

**Comment by anthonylavado** [comment]
> Expo issue linked above doesn’t look like it will ever be fixed. Is there an alternative?

@twentyh Yes? Kind of. We could eject from Expo and see about starting to re-implement this app as a “closer” to react-native project, but then that means that to work and build the iOS version, you’d have to use a Mac for development (right now that’s not required). There’s a variety of options we’re weighing internally.

**Comment by TwentyH** [comment]
> > Expo issue linked above doesn’t look like it will ever be fixed. Is there an alternative?
> 
> @TwentyH Yes? Kind of. We could eject from Expo and see about starting to re-implement this app as a “closer” to react-native project, but then that means that to work and build the iOS version, you’d have to use a Mac for development (right now that’s not required). There’s a variety of options we’re weighing internally.

Maybe I am not operating it properly, but the iOS app is unusable as it is today, anyway, pausing every time the screen locks. Wouldn’t making it a native app work around this problem, even if it is used only for streaming (UPnP or DLNA), and keep the current web view for managing the playlist?

**Comment by anthonylavado** [comment]
@TwentyH It works in limited use cases. People were clamouring for an app for iOS, but we don't have any Swift or Objective-C developers on the team. This is the best compromise we could come up with in this time.

There are other options that we are continuing to discuss internally, because we do want this experience to improve.

**Comment by kruzader** [comment]
@anthonylavado I am an iOS developer and would be interested in developing the native app as a regular user of Jellyfin. In case we could talk about it.

**Comment by dkanada** [comment]
@kruzader we can definitely talk about it :smile: since we have no extremely active iOS developers you can literally do whatever you want and if it's better than the current solution we can release it officially and start maintaining it. There is an old Swift API client in the organization you can start with if that helps at all, but it might need a few changes to get working on the latest releases.

**Comment by dkanada** [comment]
I found the client [here](https://github.com/jellyfin-archive/emby-apiclient-swift) and it looks like we never made any changes, but it should be 99% functional with the current server. I will say that based on our work with other API clients you will probably want to make a lot of changes.

**Comment by drazil100** [comment]
Was checking in on this since being able to use the jellyfin app as my primary music player is very important to me. Thought I’d also share what my current temporary solution is while I wait for background play. 

After hours of scouring the App Store and google for apps that could possibly interface with my Jellyfin server for specifically music I ended up finding foobar2000. It’s not my favorite music player I have used but it interfaces really well with UPnP and allows me to download every music album on my jellyfin server at once with album art and uses your music’s metadata to sort songs (you legit wouldn’t believe how hard it was to find an app that did all that in one package)

It’s really not an ideal solution because foobar2000’s playlists are based on file locations and therefor updating your media in anyway requires songs maintain the same file name and location on your device. 

I would REALLY like to use the jellyfin app as my primary music player and will be ditching foobar2000 the moment jellyfin is capable of being my primary music player. Just thought I’d share my temp solution for anyone in need of a more immediate solution. 

Keep up the great work guys.

**Comment by 27md** [comment]
thanks @drazil100  this works great for me until background playback is fixed

**Comment by anthonylavado** [comment]
Possibly a fix in #82, which would require testing.

**Comment by HarryPaulo** [comment]
@anthonylavado 
I tried this change #82 with "expo-keep-awake" but it didn't work.
continued to stop sound in the background with 2 Google pixel 2 and sansung A30 phones.

**Comment by anthonylavado** [comment]
@HarryPaulo Thank you for testing. The Android side of this app is... less tested (I'll leave that more to @thornbill). We do have an official Android app, as well as a few other Android things in development for music playback.

**Comment by thornbill** [comment]
I still need to test that change. I believe it only keeps the screen from automatically locking when the app is in the foreground. It’s not a real fix for playing while in the background, but it seems like a good first step. I don’t think we will get full background playback support until we implement a native player in the app.

**Comment by drazil100** [comment]
I agree its a good first step and I have often wondered about requesting this as a temporary measure. But really I just need background playback so I just didn't suggest it. 

Not having my phone fall asleep on me when I am listening to music would actually make this app usable for music at all though. At best right now I can only listen to 2 and a half minutes of music without interacting with my phone before it locks and stops background playback.

**Comment by HarryPaulo** [comment]
I have a simple radio application, I am looking for a fix for my problem with the background application.

maybe this can help other people, I will test this alternative, in my case it can solve, but I will stop using the expo-av:

https://dev.to/chafikgharbi/a-trick-to-play-sound-on-screen-lock-or-background-in-react-native-and-expo-32ml

**Comment by thornbill** [comment]
Relevant upstream issue: https://bugs.webkit.org/show_bug.cgi?id=203293

**Comment by thornbill** [comment]
There is something of a workaround for iOS 13.5. Once playback stops, you can resume playback by pressing the play button twice on the lock screen. It seems to continue playing without stopping while in the background after that. The UI does not update correctly to reflect the current state however.

**Comment by drazil100** [comment]
Just tried it and waiting for it to pause and resuming from the lock screen does in fact work. Would be nice if I didn't have to wait for it to pause itself before I can resume it (pausing and unpausing immediately does not work). Would be nicer still if it actually legitimately supported background audio playback. But I'm just about fed up with trying to sync my music library to my iPhone so maybe this is good enough until an actual fix is developed. 

Good luck to whoever eventually goes to tackle this issue.

**Comment by leinelissen** [comment]
Hey guys,

As background audio support in Expo appears to be far out and I happened to have some spare time on my hands, I opted for  creating a bare React Native app that specifically targets audio streaming, with full support for background audio and AirPlay. Find it here: [leinelissen/jellyfin-audio-player](https://github.com/leinelissen/jellyfin-audio-player). It's mostly optimised for iOS, but I figure it should work well on Android as well.

If anyone has a spare developer license lying around, I can maybe setup Testflight so trying it out is a bit easier.

Cheers!

**Comment by anthonylavado** [comment]
@leinelissen I opened an issue about it on your repository, and while we don't have a _spare_ license lying around, we do _have_ a license...

**Comment by drazil100** [comment]
A thought just occurred to me. Since getting the process to keep playing in the background is so difficult right now would it be possible to instead get it to stop immediately in a way that the workaround mentioned above can be done immediately instead of having to wait a minute for the music to stop?

Ideally of course having music just able to play in the background is the long term goal but if that’s too far in the future a short term fix might help ease the pain.

**Comment by karanbole** [comment]
> Hey guys,
> 
> As background audio support in Expo appears to be far out and I happened to have some spare time on my hands, I opted for creating a bare React Native app that specifically targets audio streaming, with full support for background audio and AirPlay. Find it here: [leinelissen/jellyfin-audio-player](https://github.com/leinelissen/jellyfin-audio-player). It's mostly optimised for iOS, but I figure it should work well on Android as well.
> 
> If anyone has a spare developer license lying around, I can maybe setup Testflight so trying it out is a bit easier.
> 
> Cheers!

I'm trying it right now, the music not playing in the background is a huge problem for this kind of audio streaming service... Looking to move away from Google play music, so Jellyfin seems like the perfect match - except this ...

**Comment by drazil100** [comment]
Same here @karanbole. I originally looked into jellyfin to get away from GPM's buggy as heck iOS app. Unfortunately jellyfin right now is video streaming first and music streaming second as far as what people seem to be focused on developing. I am still very hopeful that some of the issues like the lack of background audio playback will get fixed though especially if that music only app is any good and gets an official release. 

In the meantime I have been using foobar2000 to download music from my jellyfin server off UPnP. It's not a perfect solution but it works for my needs.

**Comment by leinelissen** [comment]
@drazil100 @karanbole If you check out the repository I mentioned you'll find there are IPA and APK packages for iOS and Android respectively. I am working on getting the app on TestFlight at least in the near future. Background audio playback is fully working. I would love to get your feedback on what it takes to create an experience you can daily drive, so that we can make audio streaming a solid part of the Jellyfin ecosystem. Feel free to try it out and leave an issue at the repository 😉

**Comment by leinelissen** [comment]
> @drazil100 @karanbole If you check out the repository I mentioned you'll find there are IPA and APK packages for iOS and Android respectively. I am working on getting the app on TestFlight at least in the near future. Background audio playback is fully working. I would love to get your feedback on what it takes to create an experience you can daily drive, so that we can make audio streaming a solid part of the Jellyfin ecosystem. Feel free to try it out and leave an issue at the repository 😉

The app is now available on TestFlight for all who want to check it out: https://testflight.apple.com/join/cf2AMDpx 🎉

**Comment by karanbole** [comment]
Great ! Installing it right now and will keep you posted 👍

**Comment by cgtobi** [comment]
@leinelissen Great job. The first tests went pretty smooth.

**Comment by atjshop** [comment]
Is this still working? I had latest iOS and it seems not working on my iPhone. Is there any other media server alternative you guys recommend that iPhone client doesn't have this limitation? Thanks.

> There is something of a workaround for iOS 13.5. Once playback stops, you can resume playback by pressing the play button twice on the lock screen. It seems to continue playing without stopping while in the background after that. The UI does not update correctly to reflect the current state however.

**Comment by the-r0cket06** [comment]
Looks like from my limited testing tonight(played 3 full songs with screen locked), this issue has been fixed with iOS 14. I believe there was an upstream bug that was finally fully fixed with iOS14.

I'd recommend anyone able to upgrade to iOS14 to do so and test.

**Comment by drazil100** [comment]
I'll have to give it a try! I hadn't even considered iOS 14 may fix it. Will report back after I test.

**Comment by drazil100** [comment]
Ok so it is confirmed to not put the music stream to sleep on iOS 14. I have no access to the next and back buttons to change songs from outside the app however. That may be better as it's own issue though.

**Comment by anthonylavado** [comment]
I can confirm the same behaviour. I'll leave it to @Thornbill to decide if this should be closed or not. Related comments are in that new issue.

**Comment by atjshop** [comment]
upgraded and confirm it is working now.

**Comment by G-eos** [comment]
work well on IOS-14 with latest Jellyfin version :) Thanks.

**Comment by thornbill** [comment]
Closing this since the issue was resolved in iOS 14! 🎉

**Comment by leocaseiro** [comment]
Hey @thornbill, is it possible to keep playing in background without the need to press "play" on lock-screen?

This feature is available in other video players such as VLC, and youtube premium.

**Comment by leocaseiro** [comment]
Ok, I was able to get the background music to work for mp3 only on IOS, but it doesn't work on videos. Is it possible to add the same feature to a video (mp4, avi, etc..)?

When I lock a screen to a playing video, it pauses automatically, and I have to press play after that. The same goes if I swap apps, from Jellyfin to another app.

**Comment by thornbill** [comment]
@leocaseiro  I would have to look into it a bit more. It may be possible to set some flags to enable background video playback.

**Comment by leocaseiro** [comment]
Hi @thornbill 
Thank you for your reply.

It's a real and known issue with expo-av, I raised a ticket and someone raised a PR in their repo.
https://github.com/expo/expo/pull/12950

**Comment by leocaseiro** [comment]
Great news! Expo just released a fix for the iOS video, now audio won't stop while video is in background.

https://github.com/expo/expo/pull/12950

I believe we need to upgrade to expo@4.2 to address this fix.

I'll try raise a PR for that, hope that won't have any breaking changes.

**Comment by thornbill** [comment]
Nice! It is a bit of a process to update Expo. There are normally some migrations that need to be done for their releases. We are a few versions behind at this point so we will need to incrementally update to 40, then 41, and finally 42 following the recommendations in their release notes listed [here](https://github.com/jellyfin/jellyfin-expo/issues/219).

**Comment by leocaseiro** [comment]
Hi, I noticed that we can only update expo-av. The only drawback is that no
longer supports iOS 10.

**Comment by thornbill** [comment]
Well it seems like this is once again an issue in iOS 15. I recommend using an alternative such as [Finamp](https://github.com/UnicornsOnLSD/finamp) in the meantime.

**Comment by drazil100** [comment]
Finamp is great! If you use Jellyfin for music you should probably just use Finamp period.

**Comment by jonatasbaldin** [comment]
I still suffer from the same problem :(

**Comment by ghost** [comment]
> Finamp is great! If you use Jellyfin for music you should probably just use Finamp period.

@drazil100 
Would do this but i would prefer only using one app as i watch movies and tv... 

~Good news! I can say that for the most part excluding this  audio playback issue the app seems to be set for iOS 16 (as of the first beta) as i have not experienced any bugs~

Edit: Never mind no external playback controls show for videos ( and of course audio) on the lock/notification screen

**Comment by nodecentral** [comment]
Just adding to this thread to say this still occurs on my iOS device , and it’s made more annoying as I use Jellyfin mainly at the gym and have to switch between apps. Would really love to see this fixed somehow so I can have uninterrupted music..

FYI - when I start the Jellyfin app up the first time it seems to work fine and is ok running in the background , but if for any reason you interact with it again and bounce between apps it risks losing that ability and will get stopped by the simplest of things e.g a pop up notification .

**Comment by janiversen** [comment]
iPad Air with iPadOS v17.1.2, newest Jellyfin-mobile from Apple Store installed, newest released jellyfin server...

but the player still stops on screen lock or when it goes in background.

Any idea when this will be solved ?

**Comment by MININT2** [comment]
i am here to say that this is still an issue 4 years later, i'm on ios and this is one of the most annoying things, along with the fact that the next track never plays unless i'm looking at the app directly while driving.
i've come to realize that this will never be fixed or worked on considering how long it's been, it is 2024.

**Comment by janiversen** [comment]
Do you use an alternate app on iOS ? 

It is a very disturbing error, but I think you are right, there are no developers interested in fixing it.

I have been looking at swiftfin, which works quite nicely for video....I am looking at extending it to support audio as well, together with jukebox functionality (I have a RPI4, and I am tired of running airplay, when it could connect directly to Jellyfin).

**Comment by zsarosi** [comment]
For me it was working till the second last iOS update (currently 17.3.1), so it was working with 17.2 or something. Some two years ago I had the same problem, but then it was fixed till now when it does not work again.

**Comment by gnattu** [comment]
For a music-centric use case, please check out [AmpFin](https://github.com/rasmuslos/AmpFin), a native implementation of the Jellyfin music player on iOS.

The problem with this Expo implementation is not trivial to resolve because simply offloading the audio playback to the system will not be sufficient. The playback queue also needs to be offloaded, and we cannot reliably use JavaScript in the WebView to manage the music playback queue because the execution of the JS inside a webview may be paused by the OS at any time in the background. A more native implementation is required to have a reliable background playback functionality.

**Comment by drazil100** [comment]
I have been using [finamp](https://github.com/jmshrv/finamp) ever since it came out and I have been liking it. Finamp seems to have a good amount of support behind it too. I might give AmpFin a try and see how I like it but I am very happy with Finamp.

**Comment by edzrol** [comment]
Had the same issue. Changed the audio setting -> max audio channels to Auto (which is the default) fixed it for me.

Hope this helps out there.

**Comment by iChristGit** [comment]
> I have been using [finamp](https://github.com/jmshrv/finamp) ever since it came out and I have been liking it. Finamp seems to have a good amount of support behind it too. I might give AmpFin a try and see how I like it but I am very happy with Finamp.

As of iOS 17.5.1 the issue has not fixed with native Jellyfin app, thanks for recommending this app, works great and gives a premium apple music like UI vibe.

**Comment by sazarkin** [comment]
Is there any recommendations/workaround to listen audiobooks?

**Comment by nodecentral** [comment]
So just to confirm,  on IOS the only current way (workaround) for someone to preserve their music continues to play in the background  if you need to bounce between apps etc. is to use a separate Jellyfin compatible music player?

Recommended players are things like finamp, or ampfin etc. 

This will be the approach until such a point that the Jellyfin app itself is updated to support it

**Comment by drazil100** [comment]
@nodecentral This is correct. But honestly I enjoy finamp so much that even IF the Jellyfin iOS app started supporting background playback I wouldn't switch back. At this point it's less of a workaround and is more just a better experience than Jellyfin app could ever be for music playback. 

That's not me disrespecting the official Jellyfin app either. 3rd party devs have just had enough time at this point to focus on developing music player specific features that would be extremely difficult if not impossible for the official app to support, or features that would make no sense outside of the context of music. 

For example one of the features in the finamp beta is that it will remember your play queue and let you resume whatever album, shuffled playlist, or whatever else you were listening to last even if you fully close and reopen the app.

**Comment by TheColin21** [comment]
Just wanted to chime in - this is a huge issue for me, especially since finamp does not support audiobooks.
The only more or less working option currently seems to be plappa which throws api errors for me frequently.

**Comment by VoxelTek** [comment]
Apparently the `expo` issue has been resolved, so could this maybe be revisited?


Also, playback seems to be even more broken for me, as going to the home screen and then returning to Jellyfin causes a period of time where the song is playing back at an incorrect speed. Additionally, pausing music playback often causes glitching, and I've found that if I quit Jellyfin after experiencing the broken pausing, occasionally it'll cause my Jellyfin server to go offline for a couple minutes. Doesn't happen with any other (non-iOS) clients (web, Android, webOS TV), just my iPad.

https://github.com/user-attachments/assets/336adf20-ca4b-4f38-b056-1f5ac901cc05

Should I make a separate issue for this? Or is it relevant to background playback?

**Comment by thornbill** [comment]
It's been longer than I realized since I shared an update here... 😅

I don't think the expo issue linked previously is relevant since it is related to the background playback of the audio for a video file. This issue is about the background playback of audio files.

The issue primarily comes down to everything managing audio playback is currently handled on the web side of things. We can think of the playback stack as being composed of 4 main components:

1. The player plugin that interfaces with the webapp. 
2. The playback queue that tracks what is being played and handles changing tracks, shuffling, etc.
3. The audio player backend that is responsible for actually playing an audio file. 
4. The audio player UI that shows a user what is playing, buttons to control playback, shows the tracks currently in the queue, etc.

Awhile back I [experimented](https://github.com/jellyfin/jellyfin-expo/pull/339) with adding a player plugin that swapped the audio player backend for expo-av but in my testing that didn't resolve any of the current issues. The next step will be rewriting all of the above components on the React Native side which will take some time.

Since we are talking about a complete rewrite of how audio playback works in the app, I would not open any new issues related to that functionality. 

I know this remains a major pain point for people using the app so it is pretty high on my list of priorities but we do need to take care of some maintenance and I'd like to unblock download support before focusing on this.

**Comment by Nolonar** [comment]
I don't have the means to test this, so maybe I'm wrong, but from what I can tell, the relevant code is in [components/AudioPlayer.js](../blob/master/components/AudioPlayer.js) (lines 21-28):

		Audio.setAudioModeAsync({
			staysActiveInBackground: true,
			interruptionModeAndroid: InterruptionModeAndroid.DoNotMix,
			playThroughEarpieceAndroid: false,
			shouldDuckAndroid: true,
			interruptionModeIOS: InterruptionModeIOS.DoNotMix,
			playsInSilentModeIOS: true
		});

According to the [Expo documentation](https://docs.expo.dev/versions/latest/sdk/audio/#audiomode), adding a `shouldPlayInBackground: true` property might just be what we need to get background audio working on iOS.

**Comment by KaiNet-X** [comment]
Also looking at expo's documentation, expo-av is deprecated and has been replaced with [expo-audio](https://docs.expo.dev/versions/latest/sdk/audio/). It's not a drop-in replacement and will need some rewriting, but it may help with this issue and potentially the bluetooth one? For background audio, rather than passing a boolean this one needs it added to app.json I believe


`
{
  "expo": {
    ...
    "ios": {
      ...
      "infoPlist": {
        ...
        "UIBackgroundModes": [
          "audio"
        ]
      }
    }
  }
}`

**Comment by thornbill** [comment]
> I don't have the means to test this, so maybe I'm wrong, but from what I can tell, the relevant code is in [components/AudioPlayer.js](../blob/master/components/AudioPlayer.js) (lines 21-28):
> 
> ```
> 	Audio.setAudioModeAsync({
> 		staysActiveInBackground: true,
> 		interruptionModeAndroid: InterruptionModeAndroid.DoNotMix,
> 		playThroughEarpieceAndroid: false,
> 		shouldDuckAndroid: true,
> 		interruptionModeIOS: InterruptionModeIOS.DoNotMix,
> 		playsInSilentModeIOS: true
> 	});
> ```
> 
> According to the [Expo documentation](https://docs.expo.dev/versions/latest/sdk/audio/#audiomode), adding a `shouldPlayInBackground: true` property might just be what we need to get background audio working on iOS.

@Nolonar  that flag is not available in the expo version we are using. I would assume if replaced the `staysActiveInBackground` flag. Also that component was an experiment in trying to resolve this issue. It’s not used by default.

> Also looking at expo's documentation, expo-av is deprecated and has been replaced with [expo-audio](https://docs.expo.dev/versions/latest/sdk/audio/). It's not a drop-in replacement and will need some rewriting, but it may help with this issue and potentially the bluetooth one? For background audio, rather than passing a boolean this one needs it added to app.json I believe
> 
> ` { "expo": { ... "ios": { ... "infoPlist": { ... "UIBackgroundModes": [ "audio" ] } } } }`

@KaiNet-X  this is already included in our `app.json` to no avail.

As I stated in https://github.com/jellyfin/jellyfin-ios/issues/19#issuecomment-2765279459, I don’t think there is any easy fix here. Our issues seem to stem from the playback queue being managed in web so I think the only solution will be to move the entire audio playback stack to the app side.

**Comment by geekwilliams** [comment]
For people looking for a workaround for audiobook playback in 2025: It is possible to use the "Copy stream url" in the jellyfin app, then use VLC to open that url and stream the audiobook in the background with the screen locked.

Wish I could help with this bug, as I don't use Jellyfin for anything besides audiobooks, and the other music/video companion-app workarounds won't work for those

**Comment by Beraliv** [comment]
Hey folks! Just to confirm that I set up new server and new iOS Jellyfin app (ie 10.11.4) and I can reproduce the background audio issue on my iPhone (iOS 26.1 (23B85))

**Comment by brandon1024** [comment]
> [@KaiNet-X](https://github.com/KaiNet-X) this is already included in our `app.json` to no avail.

Apologies for opening this discussion again, but I came across a [page in the Expo docs](https://docs.expo.dev/versions/latest/sdk/background-task/#ios-1) that seems to indicate that `processing` might also be required for background fetch to work. The package [`expo-background-task`](https://docs.expo.dev/versions/latest/sdk/background-task/) might also be needed, but I'm not sure.

I've also found a few hints in the docs that could be helpful, notably [here](https://docs.expo.dev/versions/latest/sdk/audio/#audiosetaudiomodeasyncmode):

```
// Configure audio for background playback with mixing
await setAudioModeAsync({
  playsInSilentMode: true,
  shouldPlayInBackground: true,
  interruptionMode: 'mixWithOthers'
});
```

**Comment by vern-c** [comment]
I'd like to help with a first-pass architecture/POC for reliable iOS background audio playback, and I wanted to sanity-check the direction before opening a PR.

My read of the issue history is that this should not be treated as another `UIBackgroundModes` or Expo flag fix. The app already declares `UIBackgroundModes: ["audio", "fetch"]`, and the experimental native audio path already configures `expo-av` with `staysActiveInBackground`. PR #339 was a useful experiment, but it mostly swapped the audio backend; it did not move queue ownership, Now Playing state, or remote-command handling out of Jellyfin Web.

That lines up with the maintainer notes above: when iOS backgrounds the app, WebView JavaScript is not a reliable place to own queue advancement or active playback state. Apple's media model also points toward native ownership for this class of feature: a playback audio session/background mode for continued audio, `MPNowPlayingInfoCenter` for lock screen / Control Center metadata, and `MPRemoteCommandCenter` for play/pause/next/previous/seek commands.

The product benchmark I'm aiming for is the normal iOS media behavior users see in Apple Music, VLC, Freegal, and Finamp: audio keeps playing with the screen locked, the queue advances while locked, Bluetooth/headset controls work, and the same foundation can eventually support the lock screen player and CarPlay.

A possible staged approach:

1. Add an app-side audio playback controller for audio items only. It owns the active item, stream URL, playback state, duration, position, and interruption handling.
2. Move queue ownership app-side for active audio sessions: current index, next/previous, shuffle/repeat, and automatic advancement. This is the key part for #19 because queue advancement should not depend on WebView timers/events while backgrounded.
3. Report playback start/progress/stop from the app-side controller through the Jellyfin SDK/API so server playstate remains accurate while the WebView is suspended.
4. Keep Jellyfin Web as the browsing/library UI, but have Web hand off a queue snapshot to native playback when audio starts. Native becomes authoritative while audio is active and syncs state back to Web when foregrounded.
5. Add native Now Playing metadata and remote command handlers. That would make #156 and #384 natural follow-ups, and would give #393 a real foundation instead of treating CarPlay as a separate first step.

Suggested first PR/POC:

A feature-flagged native audio controller that can play one Jellyfin audio item, continue through lock/background, and report progress/stop. I would not claim this solves #19 until a follow-up adds native queue advancement, but it should validate the foundation and make the next PR much clearer.

A few questions before I start:

- Does this staged direction match what maintainers would want, or would you prefer a different first slice?
- Should the first PR target the current Expo 46 / `expo-av` stack, or wait for the Expo upgrade work?
- Would you prefer the first PR to be a single-item playback POC, or a native queue/controller skeleton with minimal user-facing behavior?

Happy to adapt to whatever shape would be easiest to review.
